### PR TITLE
8259956: jdk.jfr.internal.ChunkInputStream#available should return the sum of remaining available bytes

### PR DESCRIPTION
**Pull request by D-D-H**
Could I have a review of this small fix?

In the current implementation, ChunkInputStream#available only returns the available size of the current stream, which is strange for the user.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8259956](https://bugs.openjdk.java.net/browse/JDK-8259956): jdk.jfr.internal.ChunkInputStream#available should return the sum of remaining available bytes

### Reviewers
 * [Erik Gahlin](https://openjdk.java.net/census#egahlin) (@egahlin - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2138/head:pull/2138`
`$ git checkout pull/2138`

### CODE REVIEW

**Comment by egahlin** [comment]
Hi Denghui, 

Looks reasonable, but I will do a proper review of this and your other PRs once everything is set for JDK 16.

Erik

**Comment by D-D-H** [comment]
> Hi Denghui,
> 
> Looks reasonable, but I will do a proper review of this and your other PRs once everything is set for JDK 16.
> 
> Erik

okay, thanks :)

**Comment by egahlin** [comment]
Not sure things should be changed. Javadoc for Inputstream::available() says:

"Returns an estimate of the number of bytes that can be read (or skipped over) from this input stream without blocking by the next invocation of a method for this input stream."

How often the stream will block depends on the underlying BufferInputStream, so it seems correct to just delegate. Purpose of the available() method is not to return the total.

**Comment by D-D-H** [comment]
> Not sure things should be changed. Javadoc for Inputstream::available() says:
> 
> "Returns an estimate of the number of bytes that can be read (or skipped over) from this input stream without blocking by the next invocation of a method for this input stream."
> 
> How often the stream will block depends on the underlying BufferInputStream, so it seems correct to just delegate. Purpose of the available() method is not to return the total.

In my humble opinion, Most users always expect this method to return the total readable size, and this is achievable for this method, although the java doc does not force the implementer to do so,
it's meaningless if this method just returns the size of the current stream which is invisible to users, based on this return value, the user can hardly do anything.
On the contrary, if the total size can be returned, the user can do further operations based on this data, such as selecting an appropriate compression algorithm or discarding, etc. (Of course this is just my idea)

what do you think?

**Comment by egahlin** [comment]
There is a method for finding the size of a recording, Recording::getSize(). It returns a long, which means it will work in cases where the recording is larger than 2 GB. I would recommend using it if you want to know the size in a reliable way.

That said, it will probably not hurt changing so that available() return the size of what is left unread on disk, potentially in memory in the future, or from an ongoing stream in a later release. 

From what I can see, Files.newInputStream(...).available() returns the full file size. This is also what java.io.FileInputStream does and BufferInputStream delegates to available() in the wrapped stream. java.io.SequenceStream on the other hand works like ChunkInputStream today.

One problem with the implementation is that you iterate over all chunks with every call to available(). It would be nice to avoid this. 

I also wonder if you considered making total a long value and check for overflow once and near the return statement?

**Comment by D-D-H** [comment]
> There is a method for finding the size of a recording, Recording::getSize(). It returns a long, which means it will work in cases where the recording is larger than 2 GB. I would recommend using it if you want to know the size in a reliable way.
> 

But if I get the data of a period of time in the past through Recoding::getStream, I cannot know its size.


> One problem with the implementation is that you iterate over all chunks with every call to available(). It would be nice to avoid this.

I used "unstreamedSize" to cache the size.

> I also wonder if you considered making total a long value and check for overflow once and near the return statement?

yes, good idea!

File: test/jdk/jdk/jfr/api/consumer/TestChunkInputStreamAvailable.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import jdk.test.lib.jfr.EventNames;
+
+public class TestChunkInputStreamAvailable {
+
+    public static void main(String[] args) throws Exception {
+        try (Recording r = new Recording()) {
+            r.enable(EventNames.JavaThreadStatistics)
+             .withPeriod(Duration.ofMillis(100));
+            r.start();
+            File repository = Path.of(System.getProperty("jdk.jfr.repository")).toFile();
```
Comment (egahlin) [inline test/jdk/jdk/jfr/api/consumer/TestChunkInputStreamAvailable.java:51]: I would prefer a simpler and more deterministic test that executes quickly and doesn't involve timing or other events (to reduce false positives).

How about this? (haven't tried it).

```
        public static void main(String... args) throws Exception {
        try (Recording r = new Recording()) {
            r.start();
            try (Recording s = new Recording()) {
                s.start(); // rotate
                s.stop(); // rotate
            }
            r.stop();
            try (InputStream is = r.getStream(null, null)) {
                int left = is.available();
                if (left != r.getSize()) {
                    String msg = "Expected IS::available() " + left;
                    msg += " to equal recording size " + r.getSize();
                    throw new Exception(msg);
                }
                while (is.read() != -1) {
                    left--;
                    int available = is.available();
                    if (available != left) {
                        String msg = "Expected IS::available() to return " + left;
                        msg += ", got " + available;
                        throw new Exception(msg);
                    }
                }
            }
        }
    }
```

Comment (D-D-H) [reply test/jdk/jdk/jfr/api/consumer/TestChunkInputStreamAvailable.java]: Make sense.
Updated.

**Review by egahlin — APPROVED** [review]

**Comment by egahlin** [comment]
/sponsor

To reduce noise on the mailing list, you may consider pushing all changes at once, or remove the hotspot-jfr label temporarily, so notifications are not sent when you are updating the PR.

Thanks
Erik

**Pull request closed without merging**